### PR TITLE
service/ptm, citra_qt: Add Play Coins setting

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Screenshot:
image

This allows setting play coins when game is running. It is a potential race condition, but should rarely be an issue, and I've left a TODO there.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4209)

### CODE REVIEW

File: src/citra_qt/configuration/configure_system.cpp [hunk 1]
```diff
@@ -293,9 +294,24 @@ void ConfigureSystem::ReadSystemSettings() {
...
 void ConfigureSystem::applyConfiguration() {
+    // apply play coin
+    u16 new_play_coin = static_cast<u16>(ui->spinBox_play_coins->value());
+    if (play_coin != new_play_coin) {
+        // archive types must be registered to set play coins
+        if (enabled)
+            Service::FS::RegisterArchiveTypes();
+        Service::PTM::Module::SetPlayCoins(new_play_coin);
+        if (enabled)
+            Service::FS::UnregisterArchiveTypes();
```
Comment (wwylele) [inline src/citra_qt/configuration/configure_system.cpp]: This chunk of code is kind of confusing without `{}`

File: src/core/hle/service/ptm/ptm.cpp [hunk 2]
```diff
@@ -136,6 +136,53 @@ void Module::Interface::CheckNew3DS(Kernel::HLERequestContext& ctx) {
     Service::PTM::CheckNew3DS(rb);
 }
 
+void WriteGameCoinData(GameCoin gamecoin_data, bool create = false) {
```
Comment (wwylele) [inline src/core/hle/service/ptm/ptm.cpp]: Mark these two functions as `static` if you don't intend to expose them

File: src/core/hle/service/ptm/ptm.cpp [hunk 3]
```diff
@@ -147,27 +194,23 @@ Module::Module() {
...
-        auto gamecoin_result =
-            Service::FS::OpenFileFromArchive(*archive_result, gamecoin_path, open_mode);
-        if (gamecoin_result.Succeeded()) {
-            auto gamecoin = std::move(gamecoin_result).Unwrap();
-            gamecoin->backend->Write(0, sizeof(GameCoin), true,
-                                     reinterpret_cast<const u8*>(&default_game_coin));
-            gamecoin->backend->Close();
-        }
+        // Create and write an empty play coin data
+        SetPlayCoins(0, true);
```
Comment (wwylele) [inline src/core/hle/service/ptm/ptm.cpp]: The default game coin has 42 coins, so if you want consistency, this should be also 42.

Comment (jroweboy) [reply src/core/hle/service/ptm/ptm.cpp]: perhaps we should set it to 300 which is the max play coins?

**Comment by jroweboy** [comment]
Can't really comment about the service side of the implementation, but I don't think this deserves its own group box. A groupbox should be for grouping multiple things together, but we'll only ever have a single thing in there. Move it just above the "regenerate console id" and you've got my approval

**Comment by wwylele** [comment]
@jroweboy note that currently the first group box has "disable when running" feature tho, so some consideration and/or code need to be made

**Comment by jroweboy** [comment]
@wwylele sounds even better that way imo. He did have a minor concern about race conditions, and avoiding the potential for race conditions entirely seems like a good plan to me :D

**Comment by wwylele** [comment]
I am going to merge if no more comments

**Pull request merged**